### ISSUE TITLE
[Bug]: Initialising the modules multiple times across test contexts causes slow test startup and logging unnecessary libraries warning

### ISSUE THREAD

**Issue by romankhomitskyi** [microsoft/playwright#30930]
### Version

1.44

### Steps to reproduce

Clone my repo `git@github.com:romankhomitskyi/module-initialization.git`
```
cd Playwright
yarn install --frozen-lockfile
yarn playwright test --project=local-chrome specs/test.spec.ts
```

you will see that faker warning message is printed multiple times even though the usage is only one

### Expected behavior

Do not initialize modules and libraries multiple times, I guess this could cause slow test start-up in large projects where ts files more than 3k + other liblaries

### Actual behavior

Loads modules multiple times
![telegram-cloud-photo-size-2-5359747465034717771-y](https://github.com/microsoft/playwright/assets/57874507/7abe093b-6d81-4d27-badd-e9f5b35dfb68)

### Environment

```shell
System:
    OS: macOS 14.4
    CPU: (10) arm64 Apple M1 Max
    Memory: 2.59 GB / 32.00 GB
  Binaries:
    Node: 20.11.0 - ~/.nvm/versions/node/v20.11.0/bin/node
    Yarn: 1.22.21 - ~/.nvm/versions/node/v20.11.0/bin/yarn
    npm: 10.2.4 - ~/.nvm/versions/node/v20.11.0/bin/npm
  IDEs:
    VSCode: 1.89.1 - /usr/local/bin/code
  Languages:
    Bash: 5.2.15 - /opt/homebrew/bin/bash
  npmPackages:
    @playwright/test: ^1.44.0 => 1.44.0
```

**Comment by pavelfeldman** [comment]
We don't load modules mode than once in a single process. I can see that you are running code outside of the test (in the describe block). We can run that code multiple times (in runner and in each worker). All in all, there is too much context in your repro, please provided a minimal repro with a single faker dependency, using npm, with declarative projects so that we could install it safely and run. Otherwise, your project structure is hard to follow.

**Comment by romankhomitskyi** [comment]
> We don't load modules mode than once in a single process. I can see that you are running code outside of the test (in the describe block). We can run that code multiple times (in runner and in each worker). All in all, there is too much context in your repro, please provided a minimal repro with a single faker dependency, using npm, with declarative projects so that we could install it safely and run. Otherwise, your project structure is hard to follow.

@pavelfeldman 
I have updated the project
https://github.com/romankhomitskyi/module-initialization

`git clone git@github.com:romankhomitskyi/module-initialization.git`

**Comment by pavelfeldman** [comment]
Ok, so what I see on the screen is:

```
[@faker-js/faker]: faker.random.words() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.words() or faker.word.words() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
[@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
```

It suggests that `words()` was called once. And then it called `word()` 10 times as requested.

**Comment by romankhomitskyi** [comment]
> Ok, so what I see on the screen is:
> 
> ```
> [@faker-js/faker]: faker.random.words() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.words() or faker.word.words() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> [@faker-js/faker]: faker.random.word() is deprecated since v8.0 and will be removed in v9.0. Please use faker.lorem.word() or faker.word.sample() instead.
> ```
> 
> It suggests that `words()` was called once. And then it called `word()` 10 times as requested.

it prints 10 times twice
also if you create a class inside describe it will create an object twice, is it ok? It affects performance, doesn't it? it's a basic example, it could be 10  complex objects creation

Is there a way to avoid this? or Am I doing it wrong?

![image](https://github.com/microsoft/playwright/assets/57874507/158b9b80-caff-4b34-8e2b-1987b6aea6c9)

**Comment by romankhomitskyi** [comment]
should I move initialization to test / test.beforeEach()?
![image](https://github.com/microsoft/playwright/assets/57874507/f0544cb6-76d2-4a0b-b838-644196cbd5ab)

**Comment by pavelfeldman** [comment]
> also if you сreate a object inside describe it will create it twice, is it ok? It affects performance, doesn't it? it's a basic example, it could be 10 complex objects creation

Yes, this code will run once in the runner process and once in the worker process. Playwright uses declarative spec structure (describes, specs, etc). Those are executed in each process and you should only add code related to the test generation there.

> Is there a way to avoid this? or Am I doing it wrong?

Yes, you can't have code outside of the test specs, hooks and fixtures.

**Comment by pavelfeldman** [comment]
Closing as working as intended, please refer to the test runner docs for how to work with the hooks and fixtures!
